Sources/Apollo/GraphQLPaginatedQueryWatcher.swift

```diff
@@ -0,0 +1,133 @@
+
+  private let client: any ApolloClientProtocol
+
+  private var initialWatcher: GraphQLQueryWatcher<Query>?
+  private var subsequentWatchers: [GraphQLQueryWatcher<Query>] = []
+
+  private let createPageQuery: CreatePageQuery
+  private let nextPageTransform: (T?, T) -> T
+
+  private var model: T? // 🚗
+  private var resultHandler: ResultHandler?
+  private var callbackQueue: DispatchQueue
+
+  /// Designated initializer
+  ///
+  /// - Parameters:
+  ///   - client: The client protocol to pass in
+  ///   - inititalCachePolicy: The preferred cache policy for the initlal page. Defaults to `returnCacheDataAndFetch`.
+  ///   - callbackQueue: The queue for response callbacks.
+  ///   - query: The query to watch
+  ///   - createPageQuery: A function which creates a new `Query` given some pagination information.
+  ///   - transform: Transforms the `Query.Data` to the intended model
+  ///   - nextPageTransform: A function which combines extant data with the new data from the next page
+  ///   - onReceiveResults: The callback function which returns changes or an error.
+  public init(
+    client: ApolloClientProtocol,
+    inititalCachePolicy: CachePolicy = .returnCacheDataAndFetch,
+    callbackQueue: DispatchQueue = .main,
+    query: Query,
+    createPageQuery: @escaping CreatePageQuery,
+    transform: @escaping (Query.Data) -> T?,
+    nextPageTransform: @escaping (T?, T) -> T,
+    onReceiveResults: @escaping (Result<T, Error>) -> Void
+  ) {
+    self.callbackQueue = callbackQueue
+    self.client = client
+    self.createPageQuery = createPageQuery
+    self.nextPageTransform = nextPageTransform
+
+    let resultHandler: ResultHandler = { [weak self] result in
+      guard let self else { return }
+      switch result {
+      case .failure(let error):
+        guard !error.wasCancelled else { return }
+        // Forward all errors aside from network cancellation errors
+        onReceiveResults(.failure(error))
+      case .success(let graphQLResult):
+        guard let data = graphQLResult.data, let transformedModel = transform(data) else { return }
+        let model = nextPageTransform(self.model, transformedModel)
+        self.model = model
+        onReceiveResults(.success(model))
+      }
+    }
+
+    self.resultHandler = resultHandler
+    initialWatcher = client.watch(
+      query: query,
+      cachePolicy: .returnCacheDataAndFetch,
+      callbackQueue: callbackQueue,
+      resultHandler: resultHandler
+    )
+  }
+
+  public func fetch() {
+    initialWatcher?.refetch()
+    cancelSubsequentWatchers()
+  }
+
+  @discardableResult
+  public func fetchNext(page: PageInfoType) -> Bool {
+    guard page.hasNextPage,
+          let nextPageQuery = createPageQuery(page),
+          let resultHandler
+    else { return false }
+
+    let nextPageWatcher = client.watch(
+      query: nextPageQuery,
+      cachePolicy: .fetchIgnoringCacheData,
```

Comment:
I think even with cursor based, this could still be an issue if we don't use `fetchIgnoringCacheData`, just less severe, as you said. This is a function of how often the data is changing out from under you.

But I also feel like basically saying that paginated queries _cannot_ use cached data is a pretty big limitation to have... 😕

Reply:
Perhaps this is best handled as a default argument? 

```swift
init(
  ...
  paginationCachePolicy: CachePolicy = .fetchIgnoringCacheData,
  ...
) { ... } 
```